src/reducers/chatReducer.js

```diff
@@ -1,50 +1,93 @@
 import {
+  CHAT_PINCODE_REQUIRED,
+  CHAT_USE_PARANOID_MODE,
+  CHAT_READY_TO_CONNECT,
+  CHAT_INIT_CONNECTION,
+  CHAT_CONNECTION_INITIATED,
+  CHAT_DISCONNECTED,
   CHAT_ADD_MESSAGE,
   CHAT_SET_USER_STATUS,
-  CHAT_CLEAR_MESSAGES,
-  CHAT_SET_USERNAME
+  CHAT_USERNAME_SET
 } from '../actions/chatActions';
 
+const PARANOID_USERNAME = ' ';
+const USERNAME_KEY = 'username';
+
 const initialState = {
+  connecting: false,
+  connected: false,
+  shouldConnect: false,
+  pincodeRequired: document && document.location.hash.endsWith('--'),
+  paranoidMode: document && document.location.hash.endsWith('---'),
+  usernameRequired: false,
   username: '',
+  previousUsername: localStorage && localStorage.getItem(USERNAME_KEY) || '',
+  status: '',
   messages: [],
   statuses: []
 };
 
 function chatReducer(state = initialState, action) {
 
+
   switch (action.type) {
 
-  case CHAT_ADD_MESSAGE:
-    return Object.assign({}, state, {
-      messages: [...state.messages, {
-        key: action.key,
-        from: action.fromUsername + action.maybeSenderId,
-        msg: action.message
-      }]
-    });
-
-  case CHAT_CLEAR_MESSAGES:
-    return Object.assign({}, state, {
-      messages: []
-    });
-
-  case CHAT_SET_USER_STATUS:
-    return Object.assign({}, state, {
-      statuses: [...state.statuses, {
-        from: action.fromUsername,
-        status: action.userStatus,
-        created: action.created
-      }]
-    });
-
-  case CHAT_SET_USERNAME:
-    return Object.assign({}, state, {
-      username: action.username
-    });
-
-  default:
-    return state;
+    case CHAT_READY_TO_CONNECT: 
+      return Object.assign({}, state, {
+        shouldConnect: true,
+        pincodeRequired: false,
+      })
+
+    case CHAT_INIT_CONNECTION:
+      return Object.assign({}, state, {
+        messages: [],
+        connecting: true,
+        connected: false,
+        shouldConnect: false,
+        pincodeRequired: false,
+        ready: false,
+        usernameRequired: !state.username,
```

Comment:
@Spetastian So `usernameRequired` is used to determine whether we should ask the user for a username?  Could call it `askForUsername`?

Reply:
Well its not a function, its state. IMO It just represent the fact that a username is required since there are none. Then what the application does with that piece of state is something else that the Reducer does not care about. 

If we were to name it e.g `askForUsername` then that name represents expectations of how that piece of state might or might not be used in a completely different part (concern) of the application stack.

Just sharing how i think about it. I can rename it if you disagree.

Reply:
I see.  The ambiguity that I'm trying to figure out how to remove is: if "a username is required" is true, does that mean that we need a username, or that we both need a username _and we don't have one (yet)_?

Reply:
This is the initial state and originally it initially checks if the username is set and if is not, it shows the the modal where the user inputs the username.

Maybe we should just skip this propert and use the username directly. Instead of `const displaySettings = showUsernameModal || usernameRequired;` we just do `const displaySettings = showUsernameModal || username;` and then use displaySettings to decide if the modal is to be shown? Because if the username is ever not set we want the user to set it. This will naturally just happen once in the beginning of the application and after that we have a username set.

Reply:
@elimisteve What do you think about that?

Reply:
@Spetastian That sounds fantastic; love the simplicity -- great idea!

Reply:
Make this tweak and I'll merge :+1: